File: src/carbonplan/widget.js
import React from "https://esm.sh/react@18";
import ReactDOM from "https://esm.sh/react-dom@18/client";
import {
  Line,
  Map,
  Raster,
  RegionPicker,
} from "https://esm.sh/@carbonplan/maps@3?deps=react@18";
import { useColormap } from "https://esm.sh/@carbonplan/colormaps@4?deps=react@18";

let bucket = "https://storage.googleapis.com/carbonplan-maps/";

let WidgetModelContext = React.createContext();

function useModelState(name) {
  let model = React.useContext(WidgetModelContext);
  let [state, setState] = React.useState(model.get(name));
  React.useEffect(() => {
    let event = `change:${name}`;
    let cb = () => setState(model.get(name));
    model.on(event, cb);
    return () => {
      model.off(event, cb);
    };
  }, [model, name]);
  return [state, (val, options) => {
    model.set(name, val, options);
    model.save_changes();
  }];
}

function App() {
  let [source] = useModelState("source");
  let [colormap_str] = useModelState("colormap");
  let [clim] = useModelState("clim");
  let [variable] = useModelState("variable");
  let [dimensions] = useModelState("dimensions");
  let [height] = useModelState("height");
  let [opacity] = useModelState("opacity");
  let [region] = useModelState("region");
  let [selector] = useModelState("selector");
  let [mode] = useModelState("mode");
  let colormap = useColormap(colormap_str, {});

  // setting this way avoids constant re-rendering
  let model = React.useContext(WidgetModelContext);
  let regionOptions = {
    setData(data) {
      model.set("data", data);
      model.save_changes();
    },
  };

  return React.createElement(
    "div",
    { style: { height } },
    React.createElement(
      Map,
      null,
      region && React.createElement(RegionPicker, null),
      React.createElement(
        Line,
        { color: "white", source: bucket + "basemaps/land", variable: "land" },
      ),
      React.createElement(
        Raster,
        {
          colormap,
          clim,
          source,
          variable,
          dimensions,
          opacity,
          regionOptions,
          selector,
          mode,
        },
      ),
    ),
  );
}

export function render({ model, el }) {
  let root = ReactDOM.createRoot(el);
  let app = React.createElement(App);
  root.render(
    React.createElement(
      WidgetModelContext.Provider,
      { value: model },
      React.createElement(App),
    ),
  );
  return () => root.unmount();
}

File: carbonplan_maps/_widget.py
import anywidget
import traitlets

import pathlib

__all__ = ["Widget"]


class Widget(anywidget.AnyWidget):
    _esm = pathlib.Path(__file__).parent / "widget.js"

    # required
    source = traitlets.Unicode(allow_none=False).tag(sync=True)
    variable = traitlets.Unicode(allow_none=False).tag(sync=True)
    dimensions = traitlets.Tuple(allow_none=False).tag(sync=True)

    # optional
    opacity = traitlets.Float(1.0).tag(sync=True)
    colormap = traitlets.Unicode("warm").tag(sync=True)
    clim = traitlets.Tuple(default_value=(-20, 30)).tag(sync=True)
    height = traitlets.Unicode("300px").tag(sync=True)
    region = traitlets.Bool(False).tag(sync=True)
    selector = traitlets.Dict().tag(sync=True)
    mode = traitlets.Unicode("texture").tag(sync=True)

    # data
    data = traitlets.Any().tag(sync=True)
